Modernize

hannesa2 opened the pull request:
* Now it build with Github actions
* most recent Android Studio 4.0
* most recent Gradle 
* most recent libraries
* api28

Please see details in commit messages

hannesa2 commented:
Now all checks are green, when you enable Github actions
Please see it here https://github.com/hannesa2/osmdroid/pull/1

<img width="570" alt="image" src="https://user-images.githubusercontent.com/3314607/83121284-65d11900-a0d2-11ea-9cac-e4fe600f7e1f.png">

hannesa2 commented:
On some test I was forced to `@Ignore` it. Everyone is invited to fix it.

Espresso test are ugly written and causes a lot of errors. 
But I don't test it in CI, it runs just the same like Travis do. (This is not perfect)

spyhunter99 commented:
first of all, awesome work. I know screwing around with gradle is painful and i appreciate the effort that went into this. I'm perfect ok with dropping gradle fury. I actually maintain it but due to changes made in gradle and the android plugin, unfortunately many of it's functions are broken with no obvious work around. Google does what they want and don't really care about backwards compatibility, unfortunately.

The whole reason we use gradle fury is really for release processes. gradle does not make it easy unfortunately. You may want to check out the release procedure wiki page to get an idea of what's going on.

So since you left no other instructions on performing release builds on this branch and publishing a snapshot, i'm going to lay out the requirements of publishing to sonatype

- javadocs jars
- source jars
- the actual artifact (aar, apk, or jar) (also on release builds, the apk has to be signed for upload to the play store)
- a populated pom file. There's a lot of requirements here and gradles build in stuff is not sufficient
- digital signatures on all of these via gpg
- hashes on all of these
- install the aar, jars, poms to local .m2 folder

Is there a profile or some special command that i don't know of to active all these things? If there's another way to get artifacts published i'm certainly willing to listen. Websites and/or services that require me to upload my private gpg key or the signing keys for the apks are a no go. 

You also broke the "site" generation capabilities, which i guess isn't a big deal. I'm not sure how many people actually use it vs just going to the javadocs or wiki directly.

Also on release builds, we run checkstyle and findbugs and a few other metrics tools against the code base. They are included with the site generation. I'd rather keep that capability than drop it entirely.

Finally, the distribution zip which we put on the github releases page is also...incomplete. I'm not sure if it's a symptom of changes made from this PR or some stupid gradle thing. 

It does build super faster, which is nice.

hannesa2 commented:
> So since you left no other instructions on performing release builds on this branch and publishing a snapshot, i'm going to lay out the requirements of publishing to sonatype
> 
>     * javadocs jars
>     * source jars

What benefit does the users have from sonatype ? A much easier distribution way is https://jitpack.io/#osmdroid/osmdroid You have to do almost nothing !
To be honest, I **never** looked into any javadocs since 20 years. The source is github, all other are archeological relicts 

>     * the actual artifact (aar, apk, or jar) (also on release builds, the apk has to be signed for upload to the play store)

I guess, this is the solution https://github.com/Triple-T/gradle-play-publisher

>     * a populated pom file. There's a lot of requirements here and gradles build in stuff is not sufficient
>     * digital signatures on all of these via gpg
>     * hashes on all of these
>     * install the aar, jars, poms to local .m2 folder

Who needs this in local maven ?
Btw, you can do this with `./gradlew publishToMavenLocal`

> 
> Is there a profile or some special command that i don't know of to active all these things? If there's another way to get artifacts published i'm certainly willing to listen. Websites and/or services that require me to upload my private gpg key or the signing keys for the apks are a no go.
> 

You can specify hiden variables in github actions. The all should done with gradle tasks

> You also broke the "site" generation capabilities, which i guess isn't a big deal. I'm not sure how many people actually use it vs just going to the javadocs or wiki directly.

No clue, what you are speaking about. A link would be helpful
 
> Also on release builds, we run checkstyle and findbugs and a few other metrics tools against the code base. They are included with the site generation. I'd rather keep that capability than drop it entirely.

Ok, checkstyle and findbugs should be performed on each pull request.
A much future way would convert to Kotlin and use ktLint. Btw current master does not check lint at all. I introduce this check and was forced to solve several issues to make it work. 
It's not only that I removed something, I changed to more common tools !

> Finally, the distribution zip which we put on the github releases page is also...incomplete. I'm not sure if it's a symptom of changes made from this PR or some stupid gradle thing.

I guess this could be the solution https://github.com/marketplace/actions/github-action-publish-binaries or a similar one

> 
> It does build super faster, which is nice.

hannesa2 commented:
Now `checkstyle` is there again. It was a pain. (and no more hidden in an other gradle plugin)

With your `checkstyle.xml` I run into
![image](https://user-images.githubusercontent.com/3314607/83223568-696caa80-a17b-11ea-8fae-0edc15d514b3.png)

I solved it pragmatically, by using a file, where I know it works.

Btw, I pointed out that codestyle is an issue too. I can format project, but then I really touch all files.

I would separate this task in an own (followup) pull request

spyhunter99 commented:
> Who needs this in local maven ?

some folks check out the osmdroid source repo, make changes for whatever reason, then test it out in their application. Easiest way is through maven local. This is documented in our guides in the wiki\

> I never looked into any javadocs since 20 years. The source is github, all other are archeological relicts

Maybe for you, but that's not the case for everyone

> site link

http://osmdroid.github.io/osmdroid/

hannesa2 commented:
Thank you for the link ! 
All about include in other project related to 

- Gradle dependency 
- Maven dependency
- Want the latest and greatest?

is described here https://jitpack.io/#osmdroid/osmdroid
Each branch and each commit can be used !

Even 
- Building from source and using the aar in your app

Which means the local maven can be archived with an own fork and usage of it, but a push is needed 
eg https://jitpack.io/#hannesa2/osmdroid
Here is the same: each branch and each commit can be used ! 

I know, it's an other way, but the result is the same

spyhunter99 commented:
@monsieurtanuki any thoughts on this? it's a significant departure from where we are now. doesn't @MKergall use jitpack for osmbonuspack? i'm a bit hesitant to make a decision on this without at least someone else offering some opinions on it

monsieurtanuki commented:
@spyhunter99 I don't understand that much about the build process in general.
But I know that I stopped working on osmdroid because I cannot currently work on osmdroid with AS 3.6.x.
Would it be better if we could keep all the previous options AND upgrade the build process? Sure!
But as long as the new build process includes a decent list of integration options (sorry if the words are not 100% relevant), that it solves the current AS 3.6 issue, that we use industry standards understandable by other apps and that we're not forced to upgrade to the ultimate latest versions of all tools (e.g. I cannot upgrade to AS4 for the moment), that's fine by me.
Of course if there are obvious issues - like if there's no way osmbonuspack can work with osmdroid anymore even when changing the "connections" - that's another story.
In short: with my limited understanding of the issue, I'm rather in favour of this heavy refactoring.

hannesa2 commented:
Please correct me, it I'm wrong
#### What's new introduced:
* lint check

#### What's currently missing and as I see needed
* findbugs check
* publish apk to Google playstore & FDroid (https://github.com/Triple-T/gradle-play-publisher but someone should do this with the credentials)
* add apk to release (To test, someone needs a token to setup)

#### What's currently missing and as I see as NOT needed
* publish to local maven (It can be done with http://osmdroid.github.io/osmdroid/)

MKergall commented:
I share most of the views of @monsieurtanuki. 
In the current situation, I can't work anymore on OBP. 
I could upgrade to AS4, if absolutely required. 

opinion on jitpack: for OBP, the move to jitpack has been made by a contributor. I have to idea how he did, but the result is really simple to use (for me, and for OBP users).

cbalster commented on osmdroid-android/src/main/java/org/osmdroid/tileprovider/MapTileProviderBasic.java:
```diff
@@ -19,204 +18,199 @@
+        if (cacheWriter != null) {
+            tileWriter = cacheWriter;
+        } else {
+            tileWriter = new SqlTileWriter();
```
related to minsdk change

cbalster commented on osmdroid-geopackage/src/main/java/org/osmdroid/gpkg/tiles/feature/GeoPackageFeatureTileProvider.java:
```diff
@@ -32,14 +30,9 @@ public GeoPackageFeatureTileProvider(ITileSource pTileSource) {
-        } else {
-            tileWriter = new SqlTileWriter();
-        }
+        tileWriter = new SqlTileWriter();
```
related to minsdk change above

cbalster commented on osmdroid-geopackage/src/main/java/org/osmdroid/gpkg/tiles/raster/GeoPackageProvider.java:
```diff
@@ -55,11 +53,7 @@ public GeoPackageProvider(final IRegisterReceiver pRegisterReceiver,
-            } else {
-                tileWriter = new SqlTileWriter();
-            }
+            tileWriter = new SqlTileWriter();
```
related to minsdk change above

hannesa2 commented on GoogleWrapperSample/build.gradle:
```diff
@@ -34,15 +34,15 @@ dependencies {
     //implementation 'com.google.android.gms:play-services:8.+'
 
-    implementation 'com.google.android.gms:play-services-maps:8.+'
+    implementation 'com.google. android.gms:play-services-maps:8.+'
```
I would remove this old stuff later on. But it doesn't hurt me.
I fixed it

hannesa2 commented on osmdroid-android/build.gradle:
```diff
@@ -13,12 +13,17 @@ android {
     }
+
+    testOptions {
+        unitTests.returnDefaultValues = true
```
Puhh, I forgot. As far as I remember it solves two tests. But I'm not sure.
You can remove it, then you see what's raises later on

hannesa2 commented on osmdroid-android/build.gradle:
```diff
@@ -3,8 +3,12 @@ apply plugin: 'com.android.library'
     compileSdkVersion = compileSdk
+    defaultConfig {
+        minSdkVersion 14
+    }
```
This was the working compromise. 
I would go strait to 21, the older software/apps can stay on current version. 😈 
You can't make everyone happy, especially when the minority is 0.001 %

cbalster replied:
But this is not your private project, but a widely used library, so tradeoffs need to be made 😉 
I think 16 is a good compromise as there is virtually no-one using devices below that and in my experience it's not much harder to support than API19 (except maybe for file-access related stuff).

hannesa2 replied:
On the other hand 14 doesn't hurt now

hannesa2 replied:
> But this is not your private project

😄 That's right, otherwise
* this PR is merged since days. 
* CI is running
* Android X is applied
* and maybe some files are in Kotlin

cbalster replied:
Just to clarify and prevent misunderstandings: I'm not a maintainer of osmdroid, just someone like you contributing from time to time whenever I stumble upon a problem or have the free time to do some cleanup or to chime in on a PR review.

hannesa2 commented on osmdroid-android/src/main/AndroidManifest.xml:
```diff
@@ -8,5 +8,8 @@
 
+    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
+	<uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+
```
This was required to make lint happy. And it makes sense !

cbalster replied:
No, it doesn't. Those permissions are only needed for some features of osmdroid. If an app doesn't use these, the library should not force the app to require those permissions! To make Lint happy, just ignore the warnings, since it's clear that it's the client apps responsibility to get these permissions if it uses e.g. the mylocationoverlay.

hannesa2 replied:
with api >= 26 it will be never used, if you don't ask user for permission.
It's an automatism, to reduce maintenance of app which uses this library.

Btw, if you really want to remove it use

```
<uses-permission
        android:name="android.permission.ACCESS_COARSE_LOCATION"
        tools:node="remove" />
```

hannesa2 replied:
Give me the right to push branches on upstream repo, then you can transform everything on your own flavor !

cbalster replied:
> with api >= 26 it will be never used, if you don't ask user for permission.
> It's an automatism, to reduce maintenance of app which uses this library.

But the app will still inherit these permissions and they will still show up e.g. in playstore, right? Or did something change here, that I'm not aware of?
If not, I still think it's unacceptable to impose these permissions to an app, even if it does not use them.

hannesa2 replied:
> But the app will still inherit these permissions and they will still show up e.g. in playstore, right? Or did something change here, that I'm not aware of?

Removed means removed

cbalster replied:
Yes, I understand what the second part does.
What I want to say is that the default should not be to just include these permissions. But that an app developer that is using the library AND is using features that require these permissions should be responsible for adding these to his own manifest.
Otherwise everyone who uses the library and does not need location access needs to explicitly remove them or has to answer questions why her app requires a permission that is not obviously needed and has serious privacy implications. I think the default should be the least permissions necessary.
This is also inline with the android sdk: If you want to use LocationManager it's your job as an app developer to make sure you get the necessary permissions....

hannesa2 commented:
This all could be only the first step to make it work, then Android X is one of the next step

hannesa2 commented:
As a next step, someone with admin rights could enable Github actions to see how CI builds "

monsieurtanuki commented:
I agree with @cbalster on the latest comments:
* I don't like permissions, therefore I wouldn't like osmdroid to impose any by default just because lint is unhappy.
* I'd like the minimum API to be as low as possible. If I remember well, 14 or 16 are good enough.

IMHO the point of this refactoring is not to have the latest SpaceX build process, it's to clean the build process with the lowest impact on the existing code and users.
@hannesa2 I imagine that can be frustrating, but doesn't that make sense?

monsieurtanuki commented:
@hannesa2 @cbalster @spyhunter99 I lost the thread guys: is there a way we can benefit from most of this pull's commits (100%, if possible) while being reasonably cautious with the existing code, permissions and API levels?
What's the next step?

hannesa2 commented:
> As a next step, someone with admin rights could enable Github actions to see how CI builds "

This could be the very next step

monsieurtanuki commented:
If I click on "Android CI", here's what I'm supposed to edit (file called `android.yml`):
```
name: Android CI

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: set up JDK 1.8
      uses: actions/setup-java@v1
      with:
        java-version: 1.8
    - name: Build with Gradle
      run: ./gradlew build
```

hannesa2 commented:
@monsieurtanuki  I appreciate your communication competence and the necessary emphatism, a lot of maintainers at Github have no such competence !

#### Concerning to open issues:
> * I don't like permissions, therefore I wouldn't like osmdroid to impose any by default just because lint is unhappy.

On the other side, I at least don't like libs, which don't work out of the box. First you have to make a,b,c, ... then you run into an error and you ask..... then you have to apply step E.
Here the topic are permissions.
I at least spent too much time with non working libs, and there is a ways to remove them 

Is this a bad or a good behavior ?

I know, simply changing it, is less work then discuss it

monsieurtanuki commented:
I see your point, but we're not talking about downloading Candy Crush for kids: I don't expect developers to understand and use osmdroid without reading at least a paragraph of instructions or sample code. For instance [this one](https://github.com/osmdroid/osmdroid/wiki/Prerequisites). Where none of the 6 permissions is mandatory.

Regarding my personal behavior towards permissions, I don't download apps with irrelevant and so-called "dangerous" permissions, such as read access to my contacts or read/write access to my SD card. I even "discarded"/uninstalled the youtube and airbnb apps for that reason. For the record, my smartphone is on API 21, therefore I'm not asked for permissions at runtime and cannot revoke them after install.

About `lint`, it's helpful, but sometimes it's wrong. Hence the `SuppressWarnings` annotation: sometimes developers know what they're doing.
And I'm convinced that we can explicitly have the warnings removed one way (the "apply fix according to `lint`" way) or the other (the specific `SuppressWarnings` way). That's possible, isn't it?

cbalster commented:
**Regarding minSdk:** I think changing it to 14 makes sense, since this is also the lowest API the supportlib/androidx support. I don't think there are close to zero actively developed apps that support anything below that. Still this is a breaking change and warrants osmdroid 7.0. Maybe one could publish a last 6.x with everything before this commit - but this is another topic and doesn't block merging this.

**Regarding permissions:** I'm still against having these (see above), but this can also be discussed later on. :)

**Regarding github-actions:** These will be active, as soon as this is merged to master. So far the workflows look good to me. The things mentioned by @spyhunter99 above like releases, javadoc etc can be added later on. In general github actions is really powerful and should be able to handle lots of these requirements with minimal effort.

**Regarding snapshot builds and general distribution of the lib:** Switching from sonatype or maybe even maven central to jitpack should fairly easy. Jitpack can automatically generate snapshot builds based on master and versions based on tags/releases. It also provides javadoc support, if it is generated in the build. See: [here](https://jitpack.io/docs/#snapshots) and [here](https://github.com/jitpack/jitpack.io).

From my point of view this should be good to go - of course the decision must be made by you and spyhunter, especially regarding the effort to change the release process.

hannesa2 commented:
> If I click on "Android CI", here's what I'm supposed to edit (file called `android.yml`)

I'm lost, you need to do this 
https://github.com/osmdroid/osmdroid/issues/1593#issuecomment-633833417

hannesa2 commented:
To avoid this endless permission discussion, I removed it.

As a blueprint to have in release a apk automatic included with Github actions, you can have a look here 
https://github.com/hannesa2/LiveEdgeDetection/releases

<img width="541" alt="image" src="https://user-images.githubusercontent.com/3314607/83976443-2ab4be00-a8fa-11ea-8790-a65c44a3098a.png">

I can help you with this, but before I want to see this merged

hannesa2 commented:
> **Regarding github-actions:** These will be active, as soon as this is merged to master. 

What's the point to wait till merge and see result **after** merge and not **now** in this pull request ?

monsieurtanuki commented:
> > If I click on "Android CI", here's what I'm supposed to edit (file called `android.yml`)
> 
> I'm lost, you need to do this
> [#1593 (comment)](https://github.com/osmdroid/osmdroid/issues/1593#issuecomment-633833417)

Oh, I cannot see this "Settings" button.

hannesa2 commented:
> > > If I click on "Android CI", here's what I'm supposed to edit (file called `android.yml`)
> > 
> > 
> > I'm lost, you need to do this
> > [#1593 (comment)](https://github.com/osmdroid/osmdroid/issues/1593#issuecomment-633833417)
> 
> Oh, I cannot see this "Settings" button.

Then you are not a admin ?

hannesa2 commented:
I only run  `git rebase -i master --autosquash` to merge some commits

monsieurtanuki commented:
> Then you are not a admin ?

github says I'm a "collaborator".

hannesa2 commented:
> > Then you are not a admin ?
> 
> github says I'm a "collaborator".

It looks like you need _"**Admin** on this repository"_

spyhunter99 commented:
i'm not an organization admin either so not sure i can help with that. I think @neilboyd or @kurtzmarc are the only ones with that permission. 

If this is merged, how do i specify the signing key to update the playstore apk?

hannesa2 commented:
> how do i specify the signing key to update the playstore apk?

@spyhunter99

https://github.com/hannesa2/LiveEdgeDetection/commit/72075bdf8ba91cfbb4d3bcb4e04290d791cffdb Here I did it by adding fake files, to be able to build in `debug` and in `release` I decrypt the real ones on CI.
I made later on, some minor changes, so `master` is the latest state.

If you have enough trust into me, I could do it. But then I specify the encrypt password and I need the `release.keystore` with all passwords.

Or I simply help you, when you do it (This is the most afford to me)

neilboyd commented:
> i'm not an organization admin either so not sure i can help with that. I think @neilboyd or @kurtzmarc are the only ones with that permission.
> 
> If this is merged, how do i specify the signing key to update the playstore apk?

@spyhunter99 you're an owner so you should be able to do everything.

GitHub Actions are awesome!

GitHub also has secrets, which should probably help with things like certificates and passwords. That allows the pipeline to use secrets without anyone seeing them.

spyhunter99 commented:
i'm not putting the signing key in version control or in ci. local only. what are the options?

hannesa2 commented:
> i'm not putting the signing key in version control or in ci. local only. what are the options?

I alread desrcibed it, but I try to describe more detailed

> Here I did it by adding fake files, to be able to build in debug and in release I decrypt the real ones on CI.

That means https://github.com/hannesa2/LiveEdgeDetection/tree/master/signing there are two 
`release.keystore` <-- a fake, just a debug keystore
`release.keystore.enc`  <-- the real, but encrypted

`decrypt.sh` encrypt it on CI, when a variable `$CRYPT_PASS` is set by CI (and only by CI)

It's done by 
`openssl aes-256-cbc -a -d -md md5 -k "$CRYPT_PASS" -in release.keystore.enc -out release.keystore`
and in `decrypt.sh` I described how to decrypt it as well

That means, you put your signing key encrypted to version control. 
If you don't trust the encryption, I'm out

hannesa2 commented:
> local only

On the other side, I don't trust system which are build locally. Everyone need a comprehensible and deterministic build.
And additionally, to put automatic git SHA1 to a `BuildConfig` variable, to know excatly what the source is

neilboyd commented:
> i'm not putting the signing key in version control or in ci. local only. what are the options?

I think I answered that question in my previous comment.

@hannesa2 don't you think GitHub Secrets would be better than encrypted keys?

neilboyd commented:
I created a temporary PR just to see how it would look - #1598

hannesa2 commented:
Yes ! Someone enabled github actions ! Now it builds

neilboyd commented:
> Yes ! Someone enabled github actions ! Now it builds
> (I only changed last commit sha1 to see if it builds)

I don't know how that works. I didn't enable Actions - I just made a branch that caused them to be triggered, and now they're also triggered on this PR.

neilboyd commented:
As far as I can see, this PR is done and you should make another one do create the release apk. If I was you I'd do that with GitHub secrets.

But I'm not involved in this project anymore, so it's up to you guys.

cbalster commented:
@spyhunter99 If you absolutely don't want to put the key/pass in a github secret, this is a pretty frictionless, one-click way of generating signed packages locally that I use in a project: https://github.com/criticalmaps/criticalmaps-android/commit/c2662eb03ad9d65eec5f249c183fc7ba31a7fbf5

hannesa2 commented:
This PR has currently 62 comments, so its easy to loose the overview. 
This is my state https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-636430000

So a guy with the power to approve  and merge (I guess @spyhunter99 ) should write here
* what prevent a merge
* what should be done in an additionally PR

and who should do it. I can't do the security related stuff

hannesa2 commented:
I pointed out that sample `osmdroid-simple-map` was not able to start
![image](https://user-images.githubusercontent.com/3314607/85219187-09e07400-b3a2-11ea-8fcf-d92513a8fc43.png)

I'll fix it

hannesa2 commented:
@neilboyd 
> If I was you I'd do that with GitHub secrets.

I do it with Github secrets !

> But I'm not involved in this project anymore, so it's up to you guys.

I was ('m) never involved in this project, to help them is a kind of "thank you for this repo" for me

neilboyd commented on .github/workflows/Android-CI-Espresso.yml:
```diff
@@ -0,0 +1,84 @@
+          name: UnitTest-report
+          path: ${{ github.workspace }}/**/build/reports/tests/testDebugUnitTest
+  Check:
+    name: Check
```
This name could be more informative

neilboyd commented:
> I do it with Github secrets !

Maybe I'm being dumb, but I don't see anywhere in the changed files where those secrets are referred to.

Maybe my previous markdown link was too subtle. Here it is expanded: https://www.google.com/search?q=GitHub+actions+Android+signing

I didn't investigate how artifacts work with GitHub, but I would expect the signed apk to be an artifact of this build, ready to upload to Google Play.
Or it could even go a step further and publish it in the pipeline.

Or maybe I misunderstood. You're talking about secrets and certificates, but I don't see anything in the pipeline that does anything with them 😕

hannesa2 commented:
@neilboyd 
https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-640257098 
I mean, I want to see progress to not waste my time

![image](https://user-images.githubusercontent.com/3314607/85683400-7a9ecd80-b6cd-11ea-93c9-7c0608bd2130.png)

Without having this [secrets](https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-649235804), I can do nothing which will work.

neilboyd commented:
So I was right. This PR doesn't need secrets. I don't know what's holding it up.

---

![image](https://user-images.githubusercontent.com/1768373/85737229-7db2b180-b6ff-11ea-8771-fe70afafd1cb.png)

neilboyd commented:
> without having this [secrets](https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-649235804), I can do nothing which will work.

(in another PR) you can create the PR with a pipeline that refers to secrets. It'll fail until someone uploads the secrets, but uploading the secrets is an easy task. That's the whole point of secrets. You can use them without knowing them.

hannesa2 commented:
Why not create the secrets now, then I can verify, if this new task works ?

But the secrets are just one part of the game. 
Someone has to encrypt the `release.keystore` and provide (upload or commit) it. 
Here I see the most danger to have it never merged (without the password, nobody can read it !), so I wait for it, till they are there ....

cbalster commented:
Isn't the whole discussion on whether to use github actions for the release
process out of scope for this PR? The point was to have a modern build
process that allows the newest APIs and Android Studio to be used and thus
unblock further contributions to the library.
The decision on how to do release builds (whether locally by someone or via
actions) and the generating snapshots can be made afterwards - different
options to go forward have already been shown here, so anything that fits
spyhunters preferred process should be able to be covered.

On Thu, 25 Jun 2020, 16:40 hannesa2 <notifications@github.com> wrote:

> Why not create the secrets now, then I can verify, if this new task works ?
>
> But the secrets are just one part of the game.
> Someone has to encrypt the release.keystore and provide (upload or
> commit) it.
> Here I see the most danger to have it never merged (without the password,
> nobody can read it !), so I wait for it, till they are there ....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-649592074>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGWNG5ISC4QVYCP7W3MGMTRYNOUTANCNFSM4NJ75BFA>
> .
>

spyhunter99 commented:
I don't really care one way or the other. As long as we can publish
release/snapshot artifacts to some kind of public maven/nexus repo and
update the demo app in the app store and have it documented (the release
process it is), then i'm fine with it.

On Thu, Jun 25, 2020 at 11:52 AM Christian Balster <notifications@github.com>
wrote:

> Isn't the whole discussion on whether to use github actions for the release
> process out of scope for this PR? The point was to have a modern build
> process that allows the newest APIs and Android Studio to be used and thus
> unblock further contributions to the library.
> The decision on how to do release builds (whether locally by someone or via
> actions) and the generating snapshots can be made afterwards - different
> options to go forward have already been shown here, so anything that fits
> spyhunters preferred process should be able to be covered.
>
> On Thu, 25 Jun 2020, 16:40 hannesa2 <notifications@github.com> wrote:
>
> > Why not create the secrets now, then I can verify, if this new task
> works ?
> >
> > But the secrets are just one part of the game.
> > Someone has to encrypt the release.keystore and provide (upload or
> > commit) it.
> > Here I see the most danger to have it never merged (without the password,
> > nobody can read it !), so I wait for it, till they are there ....
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-649592074>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACGWNG5ISC4QVYCP7W3MGMTRYNOUTANCNFSM4NJ75BFA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/osmdroid/osmdroid/pull/1594#issuecomment-649639552>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPCIGNINQKJJKSAUCOR54DRYNXFHANCNFSM4NJ75BFA>
> .
>

hannesa2 commented:
After one month no single approve, and I don't know what blocks this PR. -It looks not good-

@monsieurtanuki
> But I know that I stopped working on osmdroid because I cannot currently work on osmdroid with AS 3.6.x.

@MKergall 
> In the current situation, I can't work anymore on OBP.

Here https://github.com/hannesa2/osmbonuspack/pull/2/commits/d1ac291d20d67297f87cf88b0b5a380fd60e9ca7 you can see to use my modernized artifact.
Additionally, If you want to continue working on OBP, I'm happy to approve PRs in my fork https://github.com/osmdroid/osmdroid/pull/1594

neilboyd commented:
> But the secrets are just one part of the game.
> Someone has to encrypt the `release.keystore` and provide (upload or commit) it.
> I can neither read the secrets, nor read the `release.keystore`, but they have to be there.

keystore is also a secret:

> https://www.google.com/search?q=GitHub+actions+Android+signing

But, as @cbalster said:

> the whole discussion on whether to use github actions **for the release process** [is] out of scope for this PR

hannesa2 commented:
> > But the secrets are just one part of the game.
> > Someone has to encrypt the `release.keystore` and provide (upload or commit) it.
> 
> keystore is also a secret:
> 
> > https://www.google.com/search?q=GitHub+actions+Android+signing

Ok, I've not medium.com story, but It's similar to what I did https://medium.com/@StefMa/how-to-store-a-android-keystore-safely-on-github-actions-f0cef9413784 (4th finding in Google)

monsieurtanuki commented:
Guys, I don't know where we're going.

My feeling is that there's confusion because too many (unrelated) things are in this PR.
59 files involved...
Sometimes it's a simple refactoring in `build.gradle` files (like `implementation` instead of `compile`), sometimes it's another simple refactoring in `java` files (but swapping `import` statements or adding spaces around `=` was not an urgent thing to do, neither was the removal of the `Build.VERSION_NUMBER < 10` if statements), and sometimes it's obviously very relevant code regarding compilation - potentially with choices that need to be discussed (like the min sdk version)

@hannesa2 I think it would make sense to split this PR into simple consecutive issues, each with a simple related PR.
The first issue being something like "Modernize part I: upgrading the `build.gradle` files with the `implementation` to `compile` switch and newer version number if relevant").
The last issue being something like "Should we use secrets for the release?".

It's probably not the only way to do things but that's the only one I can think of for the moment: going step by step.

neilboyd commented:
> I think it would make sense to split this PR into simple consecutive issues, each with a simple related PR

The description neatly lists 5 points. Perhaps that could be 5 PRs.

hannesa2 commented:
I followed this idea to split it in smaller PRs. (please see the new PRs)

Now you see changes in small PRs without CI check 😄 
As I did this PR, I run from one problem into the next problem. Now you see it as standalone, and the follow up problem you don't see.
Sorry, I stopped following this .... idea.
Btw, I'll change not much here anymore, I'm (almost) done here.
 
You can split it vertical, horizontal, or in any kind you want. The amount will be the same

monsieurtanuki commented:
@hannesa2 Thanks for your attempt of PR splitting. Of course at the end it's supposed to be the same when everything is merged, but using distinct steps (when possible) makes it easier to follow and to understand.

For the sake of testing this morning I tried this PR as a patch on my Windows computer with AS 3.6.3 but it failed with an error like `This version of Android Studio cannot open this project, please retry with Android Studio 4.0 or newer.`.  Note: AS4 is not even suggested when I click on "check for updates" in AS.

monsieurtanuki commented:
I finally managed to compile with AS 3.6.3.
The trick being to play with AS and gradle version numbers:
* AS 3.4 | April 2019 | gradle 5.1.1+
* AS 3.5 | August 2019 | gradle 6.0.1+
* AS 3.6 | February 2020 | gradle 6.0.1+
* AS 4.0 | May 2020 | gradle 6.1.1

(sources: https://en.wikipedia.org/wiki/Android_Studio#Version_history, https://stackoverflow.com/questions/17727645/how-to-update-gradle-in-android-studio, https://developer.android.com/studio/releases/)

Those values impacting
* `distributionUrl=https\://services.gradle.org/distributions/gradle-*.*.*-bin.zip` in `gradle-wrapper.properties` ("gradle" version number needed)
* `dependencies { classpath "com.android.tools.build:gradle:*.*.*"}` in the main `build.gradle` ("AS" version number needed)

Going back to #1560, we see that we could not compile (before the current PR) with AS 3.6.3, but we could with AS 3.5.3.
I've just tried the AS 3.5.3 / gradle 6.0.1 version number combination and this PR seemed to work on my AS 3.6.3.
I guess that's a starting point for the step by step process; I'm about to start with #1605 and will hopefully be able to close #1560. I'll keep you posted.

monsieurtanuki commented:
I did not manage to use the splitted PRs one by one - I guess they are too interdependant.

But as said previously, the following combination works well on top of this PR (at least on my Windows AS 3.6.3):
* `distributionUrl=https\://services.gradle.org/distributions/gradle-6.0.1-bin.zip` in `gradle-wrapper.properties`
* `dependencies { classpath "com.android.tools.build:gradle:3.5.3"}` in the main `build.gradle`

I don't agree 100% with all the settings choices regarding `lint`, but it's maybe just a matter of taste.
Except in `ProjectionTest`, I don't know why the `mWidth` is changed from `600` to `978` and some tests are `@Ignore`d - don't tell me `lint` said so ;)

@hannesa2 In short, assuming you change the values of `distributionUrl` and `classpath` to those lower values, assuming that those lower values still work with AS4, and assuming you put back the correct width value in `ProjectionTest`, I'm fine with this PR.

I've lost the thread about the release/secrets discussion, but as long as this is dealt with in another issue or everybody finally agrees on the same solution, this should not prevent this PR from being merged, IMHO.

hannesa2 commented:
> For the sake of testing this morning I tried this PR as a patch on my Windows computer with AS 3.6.3 but it failed with an error like `This version of Android Studio cannot open this project, please retry with Android Studio 4.0 or newer.`. Note: AS4 is not even suggested when I click on "check for updates" in AS.

Update Android Studio itself, or the build tools ?
* Android Studio: It should work
* build tools: on this branch, there is no need anymore

Btw, when I modernize, I really modernize. Even `3.6.3` is old.

hannesa2 commented:
> I finally managed to compile with AS 3.6.3.

`git revert b569ce183598ba50c7ef6c8e5dc38c60844d282d` _would_ be the right command. (Revert AS 4.0)

You will update to AS 4.0, today or in 5 years, but I promise, you will do it !
What prevents you to do so ? Maybe you should explain this.
Why stuck on a outdated version, when you have a most recent solution ?

With this behavior, you generate a problem in the future, like you have right now here. The present is now

hannesa2 commented:
> I don't agree 100% with all the settings choices regarding `lint`, but it's maybe just a matter of taste.
> Except in `ProjectionTest`, I don't know why the `mWidth` is changed from `600` to `978` and some tests are `@Ignore`d - don't tell me `lint` said so ;)

This is the **only** interesting question !

Current `master` tests nothing, eg https://github.com/osmdroid/osmdroid/pull/1611 only JavaScript which is in my point of view **pointless** here
<img width="563" alt="image" src="https://user-images.githubusercontent.com/3314607/85976259-6fec7d00-b9da-11ea-8708-93a501334b86.png">

Back to the test:
I don't know, why magic `600` was hopefully right in the past, and now magic `978` makes the test successful.
The tests which fails randomly, I made a shortcut and simple `@Ignore` it. I was not in the mod the fix all stuff, which nobody detects the last years and broke silently.
To see it positive, now at least the remaining tests will be tested 😄 

As you @monsieurtanuki are the inventor of the test and you know more about it, maybe you can have a deeper look.
I hope, you'll find no systematic problem, and find only the past was not good.

hannesa2 commented:
> I don't agree 100% with all the settings choices regarding `lint`, but it's maybe just a matter of taste.

I total agree, there should be no exception at all !
<img width="708" alt="image" src="https://user-images.githubusercontent.com/3314607/85976935-e473eb80-b9db-11ea-9cbc-b30233cf08fd.png">

Eg, the `MissingTranslation` I can't solve it. The other I forgot the reason, remove it and try it out
But any improvement could be done later on. 

But this is not a step backwards, before there was nothing. I fixed several lint issues and now all is checked, except the ignored

monsieurtanuki commented:
The value 600 is not perfect in itself, it was just a starting point when I developed the tests. Like an old smartphone screen size in pixel.
I tried again yesterday: all the tests, including the ignored one, passed with 600. But I'm ready to try again if you see flaws in the tests.

Actually even AS doesn't feel the urge to upgrade to v4 as it does not even suggest that, at least on my 3.6.3 windows version.

Several times, on several projects, I was in the middle of coding when I agreed to upgrade AS or a library. And it broke something, and I didn't know what and why, as it's just a black box to me. Very frustrating experience.
Did I need this upgrade now? No. Did my end users need this upgrade now? No. I was coding in a relatively safe and stable environment, and that's what I like: coding.
Would it make sense to upgrade one day or the other? Sure! But when I want it, not when I'm in the middle of something important.

Unless there are numerous dangerous bugs due to this 3.5.3 parameter, I see no strong reason to force every developer to upgrade to AS4 immediately. Good enough is good enough.

Have a nice day

pasniak commented:
User here. I am on 3.6 as much as I want to upgrade to 4.0 it really feels like project on its own (I am trying to upgrade for CPU analyzer and build times; `publishToMavenLocal` of master seemed to have worked)

hannesa2 commented:
> User here. I am on 3.6 as much as I want to upgrade to 4.0 it really feels like project on its own (I am trying to upgrade for CPU analyzer and build times; `publishToMavenLocal` of master seemed to have worked)

Is this now a pro or a contra 4.0 ?

pasniak commented:
Kinda contra for now. Do you really want to force all folks \^^/ ? As much as I like easy side of contributors just getting the default AS and building, I am a bit conservative, and AS 4 is still a bit young at 4.0.0.x as default downloaded version....and they seem to be on canary 4.2 already!

As an example of upgrade's side effect: one of my AS plugins is not working (which might force rework of my app off a legacy framework supported by that plugin) so maybe sidestep an issue create a br on AS 4.0 and track master...?

kurtzmarc commented:
Hello - one of the original developers chiming in. It sounds like the end goal is to publish osm as an artifact (I'm not sure if we're doing that now). If that's the case then for a normal user it wouldn't matter if building requires 4.0 since they would just use the artifact. A build requirement like "requires AS 4.0" should only affect the active developers and contributors. For me, all this modernization is a pro. With the android ecosystem if you don't keep up, you rot and I'd hate to see this project die.

hannesa2 commented:
> Kinda contra for now. Do you really want to force all folks ^^/ ? As much as I like easy side of contributors just getting the default AS and building, I am a bit conservative, and AS 4 is still a bit young at 4.0.0.x as default downloaded version....and they seem to be on canary 4.2 already!

You forgot 4.1-beta02 

> As an example of upgrade's side effect: one of my AS plugins is not working (which might force rework of my app off a legacy framework supported by that plugin) so maybe sidestep an issue create a br on AS 4.0 and track master...?

This is the huge disadvantage of using plugins. 😄 
I made my lesson learned with plugins too. After a while no maintainer, no changes, no interest .... simply dead. Like here with the Gradle plugin. 
This is just flavored by my own experience: Use as less plugins as possible, or accept that you will become the maintainer in the future

hannesa2 commented:
> I am a bit conservative, and AS 4 is still a bit young at 4.0.0.x as default downloaded version....and they seem to be on canary 4.2 already!

There is an other point of view for 3.6 or 4.0 question: You want not to force other to upgrade... **ok, they can stay on current master, maybe for ever**.

But they should not block the present or the progress. 
I upgraded >30 projects to 4.0, and all my known problems are gone.
https://issuetracker.google.com/u/2/issues/145362242
https://issuetracker.google.com/u/2/issues/145308576
https://issuetracker.google.com/u/2/issues/144109935
https://issuetracker.google.com/issues/150511112

As @kurtzmarc  mentioned _"if you don't keep up, you rot and I'd hate to see this project die."_

TorkelV commented:
Will this be merged any time soon?

hannesa2 commented:
> Will this be merged any time soon?

Who knows ? I've the feeling, that it will happen never.

They commit direct to master and do not check pull request for a valid build that's whym after rebase this PR to recent master, it introduced new build problems 
<img width="1007" alt="image" src="https://user-images.githubusercontent.com/3314607/94330704-299dbe80-ffc7-11ea-8020-eb3149c93c09.png">

But I see it's a wasted time to fix all there problems AFTER merge them to master, to make THIS pull request error free.

hannesa2 commented:
> Will this be merged any time soon?

Sorry, the answer is never
